### ISSUE TITLE
Possible to support OPTIONS requests for CORS?

### ISSUE THREAD

**Issue by worace** [nginxinc/nginx-s3-gateway#62]
**Is your feature request related to a problem? Please describe.**

Just tried this project out for the first time and it's great. Docs were clear and easy to get started.

One issue I ran into is that it appears OPTIONS requests currently are rejected by the JS module: https://github.com/nginxinc/nginx-s3-gateway/blob/master/common/etc/nginx/include/s3gateway.js#L410-L415

It would be great to be able to enable CORS settings since that would make it easier to use this proxy directly from the browser for client-side applications.

Is it possible to add support there, along with some means to configure CORS settings? 

**Describe the solution you'd like**

**Describe alternatives you've considered**

I tried adding CORS settings via an nginx conf file under `/etc/nginx/conf.d/` but I think the JS module is taking precedence and rejecting these requests before my settings take effect.

That said, it's definitely possible I am configuring something wrong...I am by no means an expert on either NGINX or CORS.

Here is the dockerfile I was using:

```
FROM ghcr.io/nginxinc/nginx-s3-gateway/nginx-oss-s3-gateway:latest

COPY nginx.conf /etc/nginx/conf.d/default.conf
```

And this is the config file i tried:

```
server {
    echo
    listen       80;
    server_name  localhost;

    #access_log  /var/log/nginx/host.access.log  main;

    location / {
        echo "hello"
        if ($request_method = 'OPTIONS') {
           echo "method is options"
           add_header 'Access-Control-Allow-Origin' '*';
           add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS';
           #
           # Custom headers and headers various browsers *should* be OK with but aren't
           #
           add_header 'Access-Control-Allow-Headers' 'DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range';
           #
           # Tell client that this pre-flight info is valid for 20 days
           #
           add_header 'Access-Control-Max-Age' 1728000;
           add_header 'Content-Type' 'text/plain; charset=utf-8';
           add_header 'Content-Length' 0;
           return 204;
        }
        if ($request_method = 'POST') {
            add_header 'Access-Control-Allow-Origin' '*' always;
            add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS' always;
            add_header 'Access-Control-Allow-Headers' 'DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range' always;
            add_header 'Access-Control-Expose-Headers' 'Content-Length,Content-Range' always;
        }
        if ($request_method = 'GET') {
            add_header 'Access-Control-Allow-Origin' '*' always;
            add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS' always;
            add_header 'Access-Control-Allow-Headers' 'DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range' always;
            add_header 'Access-Control-Expose-Headers' 'Content-Length,Content-Range' always;
        }

        # root   /usr/share/nginx/html;
        # index  index.html index.htm;

    }

    error_page   500 502 503 504  /50x.html;
    location = /50x.html {
        root   /usr/share/nginx/html;
    }
}
```

Apologies in advance if those are way off base, but maybe someone can steer me in a better direction if this is not the right way to configure these settings.

**Additional context**

Example:

```
nginx-proxy ➸ curl -I -X OPTIONS http://localhost/example.txt
HTTP/1.1 405 Not Allowed
Server: nginx
Date: Thu, 13 Oct 2022 17:33:32 GMT
Content-Type: text/html
Content-Length: 150
Connection: keep-alive
Allow: GET, HEAD
```

**Comment by worace** [comment]
I should add that I got those NGINX cors settings from here https://enable-cors.org/server_nginx.html

**Comment by dekobon** [comment]
Hi @worace thank you for filing this issue. I know little about CORS, so please bear with me as I try to understand and work toward a solution.

If the `OPTIONS` method was enabled, what would you imagine it doing? Would it proxy the `OPTIONS` request out to S3 and then return what S3 would return? It appears that S3 does [support](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTOPTIONSobject.html) the `OPTIONS` method. However, enabling cors on a private S3 bucket does not make much sense to me.

As such, I'm guessing that we would want to enable CORS independently within the gateway. Is CORS always defined as a set of static HTTP headers that are returned on every request?

**Comment by worace** [comment]
Good questions. I am not an expert myself but I think the overview on Mozilla's site is pretty good https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS. My understanding is that when the browser is making a request to a different domain, it will often first send an `OPTIONS` request to check the CORS info, and then send the actual request you wanted to make (assuming the CORS headers that are returned are acceptable).

I think the most sophisticated version of this would allow lots of configuration, like being able to specify which domains are allowed to make CORS requests, etc. But I suspect that kind of configuration might be too cumbersome for a tool like this, so maybe a middle ground could be a simple toggle to either have no CORS (default) or to enable CORS "wide open", meaning all domains are allowed to make CORS requests to all resources. This is what the configs I showed in my example would do.

To your point about proxying the CORS behavior to S3 vs. adding it in the gateway, I agree with you -- it's probably best to handle it in the nginx plugin, if possible, rather than in the S3 layer. My understanding is that the main benefit of this gateway is the ability to add content-serving onto S3 resources which are not directly configured themselves for public content serving. So in those cases, as you said, you probably wouldn't have the CORS stuff configured on your bucket or individual S3 objects.

So it would be handy if nginx-s3-gateway had an option to enable layering that behavior on top.

**Comment by worace** [comment]
I put together a draft of one way that this might work: https://github.com/nginxinc/nginx-s3-gateway/pull/63/files

I've not worked on an NGINX module before, much less one in JS, so not sure if I have followed the best patterns there or not. 

1 question I am unsure of, is whether there is a way to combine configuration in the JS module with downstream 'standard' configuration via nginx modules -- that was what i originally tried to do, by enabling the cors settings via nginx normal configuration. But that did not seem to work, because by the time the handler reached my configuration, the requests had already been handled by the JS module.

So, if that's the case, it may be that it has to be this way -- i.e. all the CORS handling done within JS.

Also I'm afraid I didn't really understand the integration tests. Let me know if there is anything I should do there -- should I be adding a test case that tries this before and after, for example? I'm not sure how to do that yet but I could take a stab at figuring out.

**Comment by dekobon** [comment]
> I put together a draft of one way that this might work: https://github.com/nginxinc/nginx-s3-gateway/pull/63/files
> 
> I've not worked on an NGINX module before, much less one in JS, so not sure if I have followed the best patterns there or not.
> 
> 1 question I am unsure of, is whether there is a way to combine configuration in the JS module with downstream 'standard' configuration via nginx modules -- that was what i originally tried to do, by enabling the cors settings via nginx normal configuration. But that did not seem to work, because by the time the handler reached my configuration, the requests had already been handled by the JS module.
> 
> So, if that's the case, it may be that it has to be this way -- i.e. all the CORS handling done within JS.
> 
> Also I'm afraid I didn't really understand the integration tests. Let me know if there is anything I should do there -- should I be adding a test case that tries this before and after, for example? I'm not sure how to do that yet but I could take a stab at figuring out.

Once you feel like you've done all you can, I'm happy to jump in and finish it off. I will create a branch off of the main repo that will allow both of us to collaborate once you are ready.

**Comment by worace** [comment]
Sounds good @dekobon, sorry for the short delay, I'm going to be traveling for a bit. Happy for you to take that PR from where it stands if you are comfortable with that. Otherwise I'll need to take a look after a week or so when I get back.

I think my main question remaining would be whether this approach (handling all the CORS directly via javascript) is the best, or if there's any way to allow those settings to be proxied through to normal NGINX settings. I ask about the latter mostly because I suspect that's how most users might approach this issue (for example if they just search "nginx enable cors" or whatever).

But, I'm not sure the exact semantics around this JS nginx plugin and whether it's even possible to do it as I described...obviously doing it all in JS directly _does_ work, so that's at least a viable approach.

Also 1 consideration is whether we should take any extra configuration for the host name to allow access. This point I would need to research more, to find out the different ways that people actually need to configure CORS settings. But I believe a common one is to set 1 or more domains that are allowed in the CORS header. What I have now is just CORS allow "all" with `*`. But perhaps if we made that domain configurable, that would cover a better variety of config use cases? But as I said, I would have to research this a bit more to understand what is the most general purpose approach.
